query frontend, query UI: Native histogram support

rabenhorst opened the pull request:
As followup to #6032 , this PR adds native histogram support for query frontend and query UI. Additionally this PR adds native histogram support in the `ApplyCounterResetsSeriesIterator`, which is required for histogram rates.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Native histogram support for query frontend
- Native histogram support for query UI (table and graph)
- Native histogram support in `ApplyCounterResetsSeriesIterator`
- [Enable native histograms rates](https://github.com/thanos-io/thanos/pull/6071/commits/d6b359ea9d49ef972ca276403d1e37cd5f63f3c4).

## Verification

Unit tests and e2e tests.

rabenhorst commented on internal/cortex/querier/queryrange/queryrange.proto:
```diff
@@ -98,18 +98,39 @@ message PrometheusResponseQueryableSamplesStatsPerStep {
 message Sample {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/thanos-io/thanos/internal/cortex/cortexpb.LabelAdapter"];
-  cortexpb.Sample sample = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "value"];
+  double sampleValue = 2 [(gogoproto.jsontag) = "value"];
```
I understand that this is a breaking change and invalidating cache is necessary. I made the change to be closer to the prometheus common model.

fpetkovski replied:
I think this is fine as long as it's reflected in the CHANGELOG.

yeya24 commented on internal/cortex/querier/queryrange/queryrange.proto:
```diff
@@ -98,18 +98,39 @@ message PrometheusResponseQueryableSamplesStatsPerStep {
 message SampleStream {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/thanos-io/thanos/internal/cortex/cortexpb.LabelAdapter"];
   repeated cortexpb.Sample samples = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "values"];
+  repeated SampleHistogramPair histograms = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "histogram"];
```
I am not very familiar with the new histogram, so for a sample stream, we will only have either samples or histograms? Can we have them both the same time?

rabenhorst replied:
From my understanding and according to the Prometheus common code a SampleStream can have both samples and histograms: https://github.com/prometheus/common/blob/4827ac3e6ff608c6959b8a53665e715904297a69/model/value.go#L195-L226

fpetkovski replied:
Correct, it is possible for a PromQL query to return both histograms and floats. Something like `rate` will do this because it is applied independently on each series.

yeya24 commented on CHANGELOG.md:
```diff
@@ -38,6 +38,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 - [#6010](https://github.com/thanos-io/thanos/pull/6010) *: Upgrade Prometheus to v0.41.0.
 - [#5999](https://github.com/thanos-io/thanos/pull/5999) *: Upgrade Alertmanager dependency to v0.25.0.
 - [#5887](https://github.com/thanos-io/thanos/pull/5887) Tracing: Make sure rate limiting sampler is the default, as was the case in version pre-0.29.0.
+- [#6071](https://github.com/thanos-io/thanos/pull/6071) Query Frontend: *breaking :warning:* Aligned with Prometheus common model (cache reset required)
```
Can we use a more descriptive sentence about this change. Does this mean QFE can return native histograms in its response?
Btw since v0.31.0 rc is there let's update the entry location.

rabenhorst replied:
Done.

yeya24 commented on pkg/compact/downsample/downsample.go:
```diff
@@ -617,15 +618,21 @@ func (it *ApplyCounterResetsSeriesIterator) Next() chunkenc.ValueType {
 			return chunkenc.ValNone
 		}
-		if it.chks[it.i].Next() == chunkenc.ValNone {
+		it.lastValType = it.chks[it.i].Next()
```
Is this change related to this pr? Downsample/Compact side change seems not related to QFE and UI.

rabenhorst replied:
You are right. It's not directly related but it's a fix so that `rate()` works for native histograms: https://github.com/thanos-io/thanos/blob/fdeea3917591fc363a329cbe23af37c6fff0b5f0/pkg/query/iter.go#L131

I will add this information to the first comment.

GiedriusS commented:
Sorry for the delay. Is this ready for review?

rabenhorst commented:
> Sorry for the delay. Is this ready for review?

Yes it's ready and I just rebased it.

GiedriusS commented on internal/cortex/querier/queryrange/query_range.go:
```diff
@@ -657,7 +690,24 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
+				}
+			}
+
+			if len(stream.Samples) > 0 {
```
Do we need these `if`s? Maybe we can just call append instead? :thinking:

GiedriusS commented on internal/cortex/querier/queryrange/results_cache_test.go:
```diff
@@ -36,6 +37,14 @@ var (
+		Start: 1536673680 * 1e3,
+		End:   1536716898 * 1e3,
+		Step:  120 * 1e3,
+		Query: "fake_histogram",
```
Nit: if the metric is `{"fake": "histogram"}` then perhaps this query should be `{fake="histogram"}`? This doesn't change anything, just to be correct :thinking:

GiedriusS commented on test/e2e/native_histograms_test.go:
```diff
@@ -96,48 +110,263 @@ func TestWriteNativeHistograms(t *testing.T) {
+	})
+
+	vals, err := querier.SumMetrics([]string{"http_requests_total"})
+	e2emon.WithLabelMatchers(matchers.MustNewMatcher(matchers.MatchEqual, "handler", "query"))
```
This value is unused? :thinking:

rabenhorst replied:
You are tight, should be argument and I fixed it.

rabenhorst commented on pkg/queryfrontend/downsampled.go:
```diff
@@ -97,14 +98,23 @@ forLoop:
+		var ts int64
+		if len(sampleStream.Samples) > 0 {
+			ts = sampleStream.Samples[0].TimestampMs
+		} else if len(sampleStream.Histograms) > 0 {
```
Fixed that and added test cases for mix sampleStream.

rabenhorst commented on pkg/queryfrontend/downsampled.go:
```diff
@@ -97,14 +98,23 @@ forLoop:
 func minResponseTime(r queryrange.Response) int64 {
 	var res = r.(*queryrange.PrometheusResponse).Data.Result
-	if len(res) == 0 || len(res[0].Samples) == 0 {
+	if len(res) == 0 || len(res[0].Samples) == 0 && len(res[0].Histograms) == 0 {
```
Done.

GiedriusS reviewed (commented):
Please avoid rewriting history in the future with `--force` because now I'm forced to review thousands of lines again. :cry:

rabenhorst commented:
> Please avoid rewriting history in the future with `--force` because now I'm forced to review thousands of lines again. 😢

Sry for that, will avoid it in the future. I realized what I have done when it was too late already. 😞

GiedriusS commented on test/e2e/native_histograms_test.go:
```diff
@@ -38,35 +45,53 @@ func TestQueryNativeHistograms(t *testing.T) {
-			Metric: model.Metric{},
-		},
+	t.Run("query histogram using group function for testing pushdown", func(t *testing.T) {
+		queryAndAssert(t, ctx, querier.Endpoint("http"), func() string { return fmt.Sprintf("group(%v)", testHistogramMetricName) }, ts, promclient.QueryOptions{Deduplicate: true}, model.Vector{
```
I think we forgot to enable the pushdown feature flag in `querier` :smile:

rabenhorst replied:
You are right. I enabled it for this querier.

GiedriusS commented on pkg/compact/downsample/downsample.go:
```diff
@@ -674,11 +681,11 @@ func (it *ApplyCounterResetsSeriesIterator) At() (t int64, v float64) {
 
 // TODO(rabenhorst): Needs to be implemented for native histogram support.
```
Shall we remove this?

GiedriusS commented on pkg/ui/react-app/src/pages/graph/Graph.tsx:
```diff
@@ -16,7 +16,7 @@ require('jquery.flot.tooltip');
   data: {
     resultType: string;
-    result: Array<{ metric: Metric; values: [number, string][] }>;
+    result: Array<{ metric: Metric; values: [number, string][]; histograms?: [number, Histogram][] }>;
```
Let's use `SampleValue` and `SampleHistogram` types here? :thinking:

GiedriusS commented on pkg/compact/downsample/downsample_test.go:
```diff
@@ -799,6 +799,62 @@ func TestApplyCounterResetsIterator(t *testing.T) {
+			testutil.Equals(t, tcase.expected, res)
+
+			for i := range res[1:] {
+				testutil.Assert(t, res[i+1].t >= res[i].t, "sample time %v is not monotonically increasing. previous sample %v is older", res[i+1], res[i])
```
Ahh, I see, somehow missed that `i` starts from `0` and it will not go until `39` :+1: this is a nice trick.

GiedriusS reviewed (commented):
I think Cortex will be able to use this PR as an inspiration to implement native histograms on their side, right? (https://github.com/cortexproject/cortex/issues/5060)

GiedriusS reviewed (approved):
LGTM minus those small nits, I think you forgot to add DCO on your latest commit :smile:

rabenhorst commented:
> LGTM minus those small nits, I think you forgot to add DCO on your latest commit 😄

I addressed the comments and fixe DCO. 

Btw. I track progress for native histograms in https://github.com/thanos-io/thanos/issues/5907.
